File: src/JavaScriptSourceGenerator.java
// ***************************************************************************
// Copyright (c) 2013, JST/CREST DEOS project authors. All rights reserved.
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// *  Redistributions of source code must retain the above copyright notice,
//    this list of conditions and the following disclaimer.
// *  Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in the
//    documentation and/or other materials provided with the distribution.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
// TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
// **************************************************************************

//ifdef JAVA
import java.util.ArrayList;
//endif VAJA

public class JavaScriptSourceGenerator extends SourceGenerator {

	JavaScriptSourceGenerator/*constructor*/(String TargetCode, String OutputFile, int GeneratorFlag) {
		super(TargetCode, OutputFile, GeneratorFlag);
	}

	/*field*/private boolean UseLetKeyword;
	
	public  String VisitBlockJS(GtNode Node) {
		/*local*/String Code = "";
		/*local*/GtNode CurrentNode = Node;
		while(CurrentNode != null) {
			/*local*/String Statement = this.VisitNode(CurrentNode);
			if(Statement.trim().length() >0){
				Code += this.GetIndentString() + Statement + ";" + this.LineFeed;
			}
			CurrentNode = CurrentNode.NextNode;
		}
		return Code;
	}
	
	public  String VisitBlockJSWithIndent(GtNode Node) {
		/*local*/String Code = "";
		Code += "{" + this.LineFeed;
		this.Indent();
		Code += this.VisitBlockJS(Node);
		this.UnIndent();
		Code += this.GetIndentString() + "}";
		return Code;
	}

	@Override public void VisitBinaryNode(BinaryNode Node) {
		/*local*/String FuncName = Node.Token.ParsedText;
		/*local*/String Left = this.VisitNode(Node.LeftNode);
		/*local*/String Right = this.VisitNode(Node.RightNode);
		/*local*/String Source = "(" + SourceGenerator.GenerateApplyFunc2(Node.Func, FuncName, Left, Right) + ")";
		/*local*/String operator = Node.Token.ParsedText;
		if(LibGreenTea.EqualsString(operator, "/") /*&& Node.Type == Context.IntType*/ ) {
			Source = "(" + Source + " | 0)";
		}
		this.PushSourceCode(Source);
	}

	@Override public void VisitLetNode(LetNode Node) {
		/*local*/String VarName = Node.VariableName;
		/*local*/String Source = (this.UseLetKeyword ? "let " : "var ") + " " + VarName;
		if(Node.InitNode != null) {
			Node.InitNode.Evaluate(this);
			Source += " = " + this.PopSourceCode();
		}
		Source +=  ";";
		this.VisitBlockJSWithIndent(Node.BlockNode);
		this.PushSourceCode(Source + this.PopSourceCode());
	}

	@Override public void VisitIfNode(IfNode Node) {
		/*local*/String ThenBlock = this.VisitBlockJSWithIndent(Node.ThenNode);
		/*local*/String CondExpr = this.VisitNode(Node.CondExpr);
		/*local*/String Source = "if(" + CondExpr + ") " + ThenBlock;
		if(Node.ElseNode != null) {
			Source = Source + " else " + this.VisitBlockJSWithIndent(Node.ElseNode);
		}
		this.PushSourceCode(Source);
	}

	@Override public void VisitWhileNode(WhileNode Node) {
		/*local*/String LoopBody = this.VisitBlockJSWithIndent(Node.LoopBody);
		/*local*/String CondExpr = this.VisitNode(Node.CondExpr);
		this.PushSourceCode("while(" + CondExpr + ") {" + LoopBody + "}");
	}

	@Override public void VisitForNode(ForNode Node) {
		/*local*/String LoopBody = this.VisitBlockJSWithIndent(Node.LoopBody);
		/*local*/String IterExpr = this.VisitNode(Node.IterExpr);
		/*local*/String CondExpr = this.VisitNode(Node.CondExpr);
		this.PushSourceCode("for(;" + CondExpr + "; " + IterExpr + ") {" + LoopBody + "}");
	}

	@Override public void VisitDoWhileNode(DoWhileNode Node) {
		/*local*/String LoopBody = this.VisitBlockJSWithIndent(Node.LoopBody);
		/*local*/String CondExpr = this.VisitNode(Node.CondExpr);
		this.PushSourceCode("do {" + LoopBody + "}while(" + CondExpr + ");");
	}

	@Override public void VisitTryNode(TryNode Node) {
		/*local*/String Code = "try ";
		Code += this.VisitBlockJSWithIndent(Node.TryBlock);
		/*local*/LetNode Val = (/*cast*/LetNode) Node.CatchExpr;
		Code += " catch (" + Val.Type.toString() + " " + Val.VariableName + ") ";
		Code += this.VisitBlockJSWithIndent(Node.CatchBlock);
		if(Node.FinallyBlock != null) {
			Code += " finally " + this.VisitBlockJSWithIndent(Node.FinallyBlock);
		}
		this.PushSourceCode(Code);
	}

	@Override public void VisitThrowNode(ThrowNode Node) {
		/*local*/String Expr = this.VisitNode(Node.Expr);
		this.PushSourceCode("throw " + Expr);
	}

	@Override public void VisitErrorNode(ErrorNode Node) {
		/*local*/String Expr = Node.Token.ParsedText;
		this.PushSourceCode("(function() {throw new Error(\"" + Expr + "\") })()");
	}

	@Override public void GenerateFunc(GtFunc Func, ArrayList<String> NameList, GtNode Body) {
		/*local*/int ArgCount = Func.Types.length - 1;
		/*local*/String Code = "var " + Func.GetNativeFuncName() + " = (function(";
		/*local*/int i = 0;
		while(i < ArgCount) {
			if(i > 0) {
				Code = Code + ", ";
			}
			Code = Code + NameList.get(i);
			i = i + 1;
		}
		Code = Code + ") " + this.VisitBlockJSWithIndent(Body) + ");";
		this.WriteLineCode(Code);
	}

	@Override public Object Eval(GtNode Node) {
		/*local*/String ret = this.VisitBlockJS(Node);
		this.WriteLineCode(ret);
		return ret;
	}

	@Override public void InvokeMainFunc(String MainFuncName) {
		this.WriteLineCode(MainFuncName + "();");
	}
}

File: src/TypeScript/JavaScriptSourceGenerator.js
/// <reference path="LangDeps.ts" />
//  *************************************************************************** //
//  Copyright (c) 2013, JST/CRESTproject: authors: DEOS.rights: reserved: All. //
// and: Redistributionin: useand: sourceforms: binary,or: without: with //
//  modification,permitted: arethat: providedfollowing: theare: met: conditions: //
//  //
//  * of: Redistributionscode: sourceretain: mustabove: thenotice: copyright, //
//    list: thisconditions: ofthe: anddisclaimer: following. //
//  * in: Redistributionsform: binaryreproduce: mustabove: copyright: the //
//     notice,list: thisconditions: ofthe: anddisclaimer: followingthe: in //
//    and: documentation/ormaterials: otherwith: provideddistribution: the. //
//  //
// SOFTWARE: THISPROVIDED: ISTHE: BYHOLDERS: COPYRIGHTCONTRIBUTORS: AND //
//  "IS: AS"ANY: ANDOR: EXPRESSWARRANTIES: IMPLIED, INCLUDING,NOT: LIMITED: BUT //
//  TO,IMPLIED: THEOF: WARRANTIESAND: MERCHANTABILITYFOR: FITNESSPARTICULAR: A //
// ARE: DISCLAIMED: PURPOSE.NO: INSHALL: EVENTCOPYRIGHT: THEOR: HOLDER //
// BE: CONTRIBUTORSFOR: LIABLEDIRECT: ANY, INDIRECT, INCIDENTAL, SPECIAL, //
//  EXEMPLARY,CONSEQUENTIAL: DAMAGES: OR (INCLUDING,NOT: BUTTO: LIMITED, //
// OF: PROCUREMENTGOODS: SUBSTITUTESERVICES: OR;OF: USE: LOSS, DATA,PROFITS: OR; //
// BUSINESS: INTERRUPTION: OR)CAUSED: HOWEVERON: ANDTHEORY: ANYLIABILITY: OF, //
// IN: CONTRACT: WHETHER,LIABILITY: STRICT,TORT: OR (INCLUDINGOR: NEGLIGENCE //
//  OTHERWISE)IN: ARISINGWAY: ANYOF: OUTUSE: THETHIS: SOFTWARE: OF,IF: EVEN //
// OF: ADVISEDPOSSIBILITY: THESUCH: DAMAGE: OF. //
//  ************************************************************************** //
var __extends = this.__extends || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var JavaScriptSourceGenerator = (function (_super) {
    __extends(JavaScriptSourceGenerator, _super);
    function JavaScriptSourceGenerator() {
        _super.call(this, "JavaScript");
    }
    JavaScriptSourceGenerator.prototype.GenerateMethod = function (Method, NameList, Body) {
        var ArgCount = Method.Types.length - 1;
        var Code = "var " + Method.GetNativeFuncName() + " = (function(";
        var i = 0;
        while (i < ArgCount) {
            if (i > 0) {
                Code = Code + ", ";
            }
            Code = Code + NameList.get(i);
            i = i + 1;
        }
        Code = Code + ") ";
        this.VisitBlockJS(Body);
        Code += this.PopSourceCode() + ")";
        this.PushSourceCode(Code);
        this.WriteTranslatedCode(Code);
    };

    JavaScriptSourceGenerator.prototype.VisitBlockJS = function (Node) {
        this.Indent();
        var highLevelIndent = this.GetIndentString();
        _super.prototype.VisitBlock.call(this, Node);
        this.UnIndent();
        var Size = Node.CountForrowingNode();
        var Block = this.PopManyCodeAndJoin(Size, true, "\n" + highLevelIndent, ";", null);
        this.PushSourceCode("{" + Block + "\n" + this.GetIndentString() + "}");
    };

    JavaScriptSourceGenerator.prototype.VisitConstNode = function (Node) {
        this.PushSourceCode(Node.ConstValue == null ? "null" : Node.ConstValue.toString());
    };

    JavaScriptSourceGenerator.prototype.VisitUnaryNode = function (UnaryNode) {
        UnaryNode.Expr.Evaluate(this);
        this.PushSourceCode(UnaryNode.Token.ParsedText + this.PopSourceCode());
    };

    JavaScriptSourceGenerator.prototype.VisitBinaryNode = function (Node) {
        Node.RightNode.Evaluate(this);
        Node.LeftNode.Evaluate(this);
        var operator = Node.Token.ParsedText;
        if (operator.equals("/")) {
            this.PushSourceCode("((" + this.PopSourceCode() + " " + operator + " " + this.PopSourceCode() + ") | 0)");
        } else {
            this.PushSourceCode("(" + this.PopSourceCode() + " " + operator + " " + this.PopSourceCode() + ")");
        }
    };

    JavaScriptSourceGenerator.prototype.VisitNewNode = function (Node) {
        var i = 0;
        while (i < Node.Params.size()) {
            var Param = Node.Params.get(i);
            Param.Evaluate(this);
            i = i + 1;
        }
        this.PushSourceCode("new " + Node.Type.ShortClassName + "()");
    };

    JavaScriptSourceGenerator.prototype.VisitNullNode = function (Node) {
        this.PushSourceCode("null");
    };

    JavaScriptSourceGenerator.prototype.VisitLocalNode = function (Node) {
        this.PushSourceCode(Node.LocalName);
    };

    JavaScriptSourceGenerator.prototype.VisitGetterNode = function (Node) {
        Node.Expr.Evaluate(this);
        this.PushSourceCode(this.PopSourceCode() + "." + Node.Token.ParsedText);
    };

    JavaScriptSourceGenerator.prototype.VisitIndexerNode = function (Node) {
        Node.IndexAt.Evaluate(this);
        Node.Expr.Evaluate(this);
        this.PushSourceCode(this.PopSourceCode() + "." + Node.Token.ParsedText + "[" + this.PopSourceCode() + "]");
    };

    JavaScriptSourceGenerator.prototype.VisitApplyNode = function (Node) {
        var methodName = Node.Method.GetNativeFuncName();
        var ParamCount = Node.Params.size();
        var i = 0;
        while (i < ParamCount) {
            Node.Params.get(i).Evaluate(this);
            i = i + 1;
        }
        var params = "(" + this.PopManyCodeAndJoin(ParamCount, true, null, null, ", ") + ")";
        this.PushSourceCode(methodName + params);
    };

    JavaScriptSourceGenerator.prototype.VisitAndNode = function (Node) {
        Node.LeftNode.Evaluate(this);
        Node.RightNode.Evaluate(this);
        var Right = this.PopSourceCode();
        var Left = this.PopSourceCode();
        this.PushSourceCode(Left + " && " + Right);
    };

    JavaScriptSourceGenerator.prototype.VisitOrNode = function (Node) {
        Node.LeftNode.Evaluate(this);
        Node.RightNode.Evaluate(this);
        var Right = this.PopSourceCode();
        var Left = this.PopSourceCode();
        this.PushSourceCode(Left + " || " + Right);
    };

    JavaScriptSourceGenerator.prototype.VisitAssignNode = function (Node) {
        Node.LeftNode.Evaluate(this);
        Node.RightNode.Evaluate(this);
        var Right = this.PopSourceCode();
        var Left = this.PopSourceCode();
        this.PushSourceCode((this.UseLetKeyword ? "let " : "var ") + Left + " = " + Right);
    };

    JavaScriptSourceGenerator.prototype.VisitLetNode = function (Node) {
        var VarName = Node.VariableName;
        var Source = (this.UseLetKeyword ? "let " : "var ") + " " + VarName;
        if (Node.InitNode != null) {
            Node.InitNode.Evaluate(this);
            Source += " = " + this.PopSourceCode();
        }
        Source += ";";
        this.VisitBlockJS(Node.BlockNode);
        this.PushSourceCode(Source + this.PopSourceCode());
    };

    JavaScriptSourceGenerator.prototype.VisitIfNode = function (Node) {
        Node.CondExpr.Evaluate(this);
        this.VisitBlockJS(Node.ThenNode);
        var ThenBlock = this.PopSourceCode();
        var CondExpr = this.PopSourceCode();
        var Source = "if(" + CondExpr + ") " + ThenBlock;
        if (Node.ElseNode != null) {
            this.VisitBlockJS(Node.ElseNode);
            Source = Source + " else " + this.PopSourceCode();
        }
        this.PushSourceCode(Source);
    };

    JavaScriptSourceGenerator.prototype.VisitSwitchNode = function (Node) {
        // Auto: TODO-generatedstub: method //
    };

    JavaScriptSourceGenerator.prototype.VisitWhileNode = function (Node) {
        Node.CondExpr.Evaluate(this);
        this.VisitBlockJS(Node.LoopBody);
        var LoopBody = this.PopSourceCode();
        var CondExpr = this.PopSourceCode();
        this.PushSourceCode("while(" + CondExpr + ") {" + LoopBody + "}");
    };

    JavaScriptSourceGenerator.prototype.VisitForNode = function (Node) {
        Node.CondExpr.Evaluate(this);
        Node.IterExpr.Evaluate(this);
        this.VisitBlockJS(Node.LoopBody);
        var LoopBody = this.PopSourceCode();
        var IterExpr = this.PopSourceCode();
        var CondExpr = this.PopSourceCode();
        this.PushSourceCode("for(;" + CondExpr + "; " + IterExpr + ") {" + LoopBody + "}");
    };

    JavaScriptSourceGenerator.prototype.VisitDoWhileNode = function (Node) {
        Node.CondExpr.Evaluate(this);
        this.VisitBlockJS(Node.LoopBody);
        var LoopBody = this.PopSourceCode();
        var CondExpr = this.PopSourceCode();
        this.PushSourceCode("do {" + LoopBody + "}while(" + CondExpr + ");");
    };

    JavaScriptSourceGenerator.prototype.VisitForEachNode = function (ForEachNode) {
        // Auto: TODO-generatedstub: method //
    };

    JavaScriptSourceGenerator.prototype.VisitEmptyNode = function (Node) {
        this.PushSourceCode("");
    };

    JavaScriptSourceGenerator.prototype.VisitReturnNode = function (Node) {
        if (Node.Expr != null) {
            Node.Expr.Evaluate(this);
            this.PushSourceCode("return " + this.PopSourceCode());
        } else {
            this.PushSourceCode("return");
        }
    };

    JavaScriptSourceGenerator.prototype.VisitBreakNode = function (Node) {
        this.PushSourceCode("break");
    };

    JavaScriptSourceGenerator.prototype.VisitContinueNode = function (Node) {
        this.PushSourceCode("continue");
    };

    JavaScriptSourceGenerator.prototype.VisitTryNode = function (Node) {
        this.VisitBlockJS(Node.TryBlock);

        // 		/* FIXME:not: Dofor: statement: use */for(var i: number = 0; i < Node.CatchBlock.size(); i++) { //
        // 			var Block: TypedNode = (TypedNode) Node.CatchBlock.get(i); //
        // 			var Exception: TypedNode = (TypedNode) Node.TargetException.get(i); //
        // 			this.VisitBlockJS(Block); //
        // 		} //
        this.VisitBlockJS(Node.FinallyBlock);

        var FinallyBlock = this.PopSourceCode();
        var CatchBlocks = "";
        var TryBlock = this.PopSourceCode();
        this.PushSourceCode("try " + TryBlock + "finally " + FinallyBlock);
        return;
    };

    JavaScriptSourceGenerator.prototype.VisitThrowNode = function (Node) {
        Node.Expr.Evaluate(this);
        var Expr = this.PopSourceCode();
        this.PushSourceCode("throw " + Expr);
        return;
    };

    JavaScriptSourceGenerator.prototype.VisitFunctionNode = function (Node) {
        // Auto: TODO-generatedstub: method //
        return;
    };

    JavaScriptSourceGenerator.prototype.VisitErrorNode = function (Node) {
        var Expr = Node.toString();
        this.PushSourceCode("(function() {throw new Error(\"" + Expr + "\") })()");
        return;
    };

    // must: Thisextended: beeach: language: in //
    JavaScriptSourceGenerator.prototype.Eval = function (Node) {
        this.VisitBlock(Node);
        var ret = "";
        while (this.GeneratedCodeStack.size() > 0) {
            var Line = this.PopSourceCode();
            if (Line.length > 0) {
                ret = Line + ";\n" + ret;
            }
        }

        // this.WriteTranslatedCode(ret); //
        return ret;
    };
    return JavaScriptSourceGenerator;
})(SourceGenerator);
